ios/keyman/Keyman/Keyman/KMWebBrowser/WebBrowserViewController.swift

```diff
@@ -204,6 +204,38 @@ class WebBrowserViewController: UIViewController, UIWebViewDelegate, UIAlertView
   func webView(_ webView: UIWebView,
                shouldStartLoadWith request: URLRequest,
                navigationType: UIWebView.NavigationType) -> Bool {
+    if request.url?.lastPathComponent.hasSuffix(".kmp") ?? false {
```

Comment:
This may not be 100% correct. We should be looking at the [content-disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) header if it exists to determine the filename. We cannot assume that a URL will have a .kmp extension (e.g. it might be called something like `file-downloader.php`)

Reply:
When using the keyman.com language search. the initial link _is_ PHP, though that ends up resolving to a .kmp that this can then intercept.

I'm _completely_ unfamiliar with such headers; have any good example ones you could point me at for this?

Reply:
Like, for an actual .kmp.

Reply:
I don't have one handy, but you could throw this onto a local website together with khmer_angkor.kmp (which you can even rename if you want, e.g. in the script below I renamed the copy I am reading from to example_khmer_angkor.dat just to illustrate). You could save this as `download.php`.

```php
<?php
  $file = './example_khmer_angkor.dat';
  header('Content-Type: application/octet-stream');
  header('Content-Disposition: attachment; filename=khmer_angkor.kmp');
  header('Content-Length: ' . filesize($file));
  readfile($file);
```

Reply:
I'll test with that shortly, but first I wanted to see what I can get from the keyboard search.

From the [keyboard page for `sil_ipa`](https://keyman.com/keyboards/sil_ipa), the link seems to process in three stages:

Stage 1:

```
▿ https://keyman.com/keyboard/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
  ▿ url : Optional<URL>
    ▿ some : https://keyman.com/keyboard/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
      - _url : https://keyman.com/keyboard/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
  - cachePolicy : 0
  - timeoutInterval : 2147483647.0
  ▿ mainDocumentURL : Optional<URL>
    ▿ some : https://keyman.com/keyboard/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
      - _url : https://keyman.com/keyboard/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
  - networkServiceType : __C.NSURLRequestNetworkServiceType
  - allowsCellularAccess : true
  ▿ httpMethod : Optional<String>
    - some : "GET"
  ▿ allHTTPHeaderFields : Optional<Dictionary<String, String>>
    ▿ some : 3 elements
      ▿ 0 : 2 elements
        - key : "Accept"
        - value : "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
      ▿ 1 : 2 elements
        - key : "User-Agent"
        - value : "Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148"
      ▿ 2 : 2 elements
        - key : "Referer"
        - value : "https://keyman.com/keyboards/sil_ipa"
  - httpBody : nil
  - httpBodyStream : nil
  - httpShouldHandleCookies : true
  - httpShouldUsePipelining : true
``` 

Stage 2:

```
▿ https://keyman.com/keyboards/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
  ▿ url : Optional<URL>
    ▿ some : https://keyman.com/keyboards/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
      - _url : https://keyman.com/keyboards/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
  - cachePolicy : 0
  - timeoutInterval : 2147483647.0
  ▿ mainDocumentURL : Optional<URL>
    ▿ some : https://keyman.com/keyboards/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
      - _url : https://keyman.com/keyboards/download?id=sil_ipa&platform=ios&mode=standalone&cid=760108319.1607477546
  - networkServiceType : __C.NSURLRequestNetworkServiceType
  - allowsCellularAccess : true
  ▿ httpMethod : Optional<String>
    - some : "GET"
  ▿ allHTTPHeaderFields : Optional<Dictionary<String, String>>
    ▿ some : 5 elements
      ▿ 0 : 2 elements
        - key : "Accept-Encoding"
        - value : "gzip, deflate, br"
      ▿ 1 : 2 elements
        - key : "Accept"
        - value : "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
      ▿ 2 : 2 elements
        - key : "Referer"
        - value : "https://keyman.com/keyboards/sil_ipa"
      ▿ 3 : 2 elements
        - key : "User-Agent"
        - value : "Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148"
      ▿ 4 : 2 elements
        - key : "Accept-Language"
        - value : "en-us"
  - httpBody : nil
  - httpBodyStream : nil
  - httpShouldHandleCookies : true
  - httpShouldUsePipelining : true
```

Stage 3:

```
▿ https://downloads.keyman.com/keyboards/sil_ipa/1.8.4/sil_ipa.kmp
  ▿ url : Optional<URL>
    ▿ some : https://downloads.keyman.com/keyboards/sil_ipa/1.8.4/sil_ipa.kmp
      - _url : https://downloads.keyman.com/keyboards/sil_ipa/1.8.4/sil_ipa.kmp
  - cachePolicy : 0
  - timeoutInterval : 2147483647.0
  ▿ mainDocumentURL : Optional<URL>
    ▿ some : https://downloads.keyman.com/keyboards/sil_ipa/1.8.4/sil_ipa.kmp
      - _url : https://downloads.keyman.com/keyboards/sil_ipa/1.8.4/sil_ipa.kmp
  - networkServiceType : __C.NSURLRequestNetworkServiceType
  - allowsCellularAccess : true
  ▿ httpMethod : Optional<String>
    - some : "GET"
  ▿ allHTTPHeaderFields : Optional<Dictionary<String, String>>
    ▿ some : 5 elements
      ▿ 0 : 2 elements
        - key : "Accept-Encoding"
        - value : "gzip, deflate, br"
      ▿ 1 : 2 elements
        - key : "Accept-Language"
        - value : "en-us"
      ▿ 2 : 2 elements
        - key : "Referer"
        - value : "https://keyman.com/keyboards/sil_ipa"
      ▿ 3 : 2 elements
        - key : "Accept"
        - value : "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
      ▿ 4 : 2 elements
        - key : "User-Agent"
        - value : "Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148"
  - httpBody : nil
  - httpBodyStream : nil
  - httpShouldHandleCookies : true
  - httpShouldUsePipelining : true
```

Reply:
Setting that up under my local keymanweb.com development 'server' setup...

```
▿ http://keymanweb.com.local/test.php
  ▿ url : Optional<URL>
    ▿ some : http://keymanweb.com.local/test.php
      - _url : http://keymanweb.com.local/test.php
  - cachePolicy : 0
  - timeoutInterval : 60.0
  ▿ mainDocumentURL : Optional<URL>
    ▿ some : http://keymanweb.com.local/test.php
      - _url : http://keymanweb.com.local/test.php
  - networkServiceType : __C.NSURLRequestNetworkServiceType
  - allowsCellularAccess : true
  ▿ httpMethod : Optional<String>
    - some : "GET"
  ▿ allHTTPHeaderFields : Optional<Dictionary<String, String>>
    ▿ some : 2 elements
      ▿ 0 : 2 elements
        - key : "Accept"
        - value : "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
      ▿ 1 : 2 elements
        - key : "User-Agent"
        - value : "Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148"
  - httpBody : nil
  - httpBodyStream : nil
  - httpShouldHandleCookies : true
  - httpShouldUsePipelining : true
```

There are no further stages; the next thing that occurs is a "Frame load interrupted" error.  So that's _all_ iOS will give me without _very_ significant changes, unless I'm missing something.

Reply:
Just to make clear the testing setup I used:

<img width="1136" alt="Screen Shot 2021-03-04 at 2 59 46 PM" src="https://user-images.githubusercontent.com/25213402/109930505-508cb900-7cfa-11eb-80dd-4181de7376d5.png">